### PR TITLE
Extend bridge info support

### PR DESCRIPTION
**Pull request by ronand-atl**
This PR adds a few more supported attributes in route/link/bridge_info.

Support for the following bridge attributes is added:
- IFLA_BR_STP_STATE
- IFLA_BR_MCAST_ROUTER
- IFLA_BR_MCAST_SNOOPING

Support for IFLA_BR_MULTI_BOOLOPT is added, so the following boolopts can be set:
- BR_BOOLOPT_NO_LL_LEARN
- BR_BOOLOPT_MCAST_VLAN_SNOOPING
- BR_BOOLOPT_MST_ENABLE

Additionally, I've added a few missing getter functions for already supported attributes.

### CODE REVIEW

File: libnl-route-3.sym [hunk 1]
```diff
@@ -1340,5 +1340,9 @@ global:
 	rtnl_link_bond_get_miimon;
 	rtnl_link_bond_get_min_links;
 	rtnl_link_bond_get_mode;
+	rtnl_link_bridge_get_nf_call_arptables;
+	rtnl_link_bridge_get_nf_call_ip6tables;
+	rtnl_link_bridge_get_nf_call_iptables;
+	rtnl_link_bridge_set_vlan_default_pvid;
```
Comment (thom311) [inline libnl-route-3.sym:1358]: Optimally, this bugfix would be a separate commit. Could you be bothered to split it out? :)

File: include/netlink/route/link/bridge_info.h [hunk 2]
```diff
@@ -6,6 +6,7 @@
 #ifndef NETLINK_LINK_BRIDGE_INFO_H_
 #define NETLINK_LINK_BRIDGE_INFO_H_
 
+#include <stdbool.h>
```
Comment (thom311) [inline include/netlink/route/link/bridge_info.h:9]: `<stdbool.h>` is a C99 header. So far we don't use that in public headers.

However, we already use `<stdint.h>` (also C99), so we can assume that all users use C99 and will also have <stdbool.h>.

One concern is that `<stdbool.h>` redefines `bool`/`true`/`false`. So if a user had their own such identifiers, this change will break them (well, at least, if they use this particular header -- which probably means that nobody is going to be affected). However, this also creates a precedence, so soon we might use `<stdbool.h>` in more prominent headers, where it might break some users.

Unclear what to do about that. Maybe just use an `int`?

File: include/netlink/route/link/bridge_info.h [hunk 3]
```diff
@@ -68,6 +69,19 @@ extern void rtnl_link_bridge_set_mcast_snooping(struct rtnl_link *link,
 extern int rtnl_link_bridge_get_mcast_snooping(struct rtnl_link *link,
 					       uint8_t *value);
 
+extern void rtnl_link_bridge_set_no_linklocal_learn(struct rtnl_link *link,
+						    bool value);
+extern int rtnl_link_bridge_get_no_linklocal_learn(struct rtnl_link *link,
+						   bool *value);
```
Comment (thom311) [inline include/netlink/route/link/bridge_info.h:75]: In this case, we can return the boolean value and an error value at the same time.

In earlier places, where we get a `uint32_t`, a `int` cannot encode a (negative) error value and the return value. So we had the output argument. I think for getters where successful values fit into `[0, INT_MAX]`, we can fit them both together. It's more convenient to use.

Consistency is a concern here, however, we are already not quite consistent (sometimes for good reasons). So, it seems more useful to me to use the API that makes the most sense for the type. Meaning, a getter for a boolean can squeeze the error and the boolean value in an `int`.

File: lib/route/link/bridge_info.c [hunk 4]
```diff
@@ -239,6 +256,55 @@ static struct rtnl_link_info_ops bridge_info_ops = {
...
+
+	mask = 1ul << opt;
+
+	if (value)
+		bi->b_boolopts.optval |= mask;
+	else
+		bi->b_boolopts.optval &= ~mask;
+
+	bi->b_boolopts.optmask |= mask;
+	bi->ce_mask |= BRIDGE_ATTR_BOOLOPT;
```
Comment (thom311) [inline lib/route/link/bridge_info.c:278]: There is no API to unset `bi->ce_mask & BRIDGE_ATTR_BOOLOPT` or `bi->b_boolopts.optmask & mask`.

That's a common lack in many setters, so it's probably fine. 

If we make `value` an `int`, we could take "-1" to unset `bi->b_boolopts.optmask`. Not sure that is useful though. It's also slightly confusing. So maybe not.

Fine-as-is, unless you'd like to change it.

File: lib/route/link/bridge_info.c [hunk 5]
```diff
@@ -779,6 +845,99 @@ int rtnl_link_bridge_get_mcast_snooping(struct rtnl_link *link, uint8_t *value)
...
+ * @arg link	Link object of type bridge
+ * @arg value	Output argument.
+ *
+ * @see rtnl_link_bridge_set_no_linklocal_learn()
+ *
+ * @return Zero on success, otherwise a negative error code.
+ * @retval -NLE_NOATTR
+ * @retval -NLE_INVAL
+ */
+int rtnl_link_bridge_get_no_linklocal_learn(struct rtnl_link *link, bool *value)
```
Comment (thom311) [inline lib/route/link/bridge_info.c:873]: I really think that instead of adding all those `rtnl_link_bridge_get_no_linklocal_learn()` accessors, just expose `bridge_info_set_boolopt()` as public API.

Note that enum values like `BR_BOOLOPT_NO_LL_LEARN` are already public API (in kernel headers). Just let the user include `if_bridge.h` and set the flags themselves. The benefit is, that we support also future enum values.

Also, no need for the check `if (opt < 0 || opt >= BR_BOOLOPT_MAX)`, if the user sets a value that is not supported then kernel will tell us. On the other hand, we must check that opt is < 32, so that the bitshift is not undefined behavior.

**Review by thom311 — CHANGES_REQUESTED** [review]

File: libnl-route-3.sym (hunk 1 above)
Comment (ronand-atl) [reply libnl-route-3.sym:1354]: Sure

File: include/netlink/route/link/bridge_info.h (hunk 2 above)
Comment (ronand-atl) [reply include/netlink/route/link/bridge_info.h]: Personally I think that using an int over a bool for new code would probably be a step backwards. However, since your suggestion lines up nicely with your other suggestion, I will change it to use an int.

File: lib/route/link/bridge_info.c (hunk 4 above)
Comment (ronand-atl) [reply lib/route/link/bridge_info.c]: I also noticed the common lack of a way to unset values. I'm not sure of the usefulness of that though, since you could just not set the value in the first place. Maybe it would be useful for getting an object, modifying it a bit, and sending it back as a change request which matches a different object. That seems like an odd use case though. I think I'll leave it as is for now. If there is a desire in the future for ways to unset values, maybe a `reset` or `clear` version of the functions could be added.

File: lib/route/link/bridge_info.c (hunk 5 above)
Comment (ronand-atl) [reply lib/route/link/bridge_info.c]: I thought that might be brought up. I initially did it that way because I thought it might be nice to have a higher level API that abstracts the boolopts detail.

I've made the boolopts setter and getter public and removed the individual functions for now.

File: include/netlink/route/link/bridge_info.h [hunk 6]
```diff
@@ -8,6 +8,7 @@
 
 #include <netlink/netlink.h>
 #include <netlink/route/link.h>
+#include <linux/if_bridge.h>
```
Comment (ronand-atl) [inline include/netlink/route/link/bridge_info.h:11]: This probably won't fly. Are there any other solutions other than just using an int instead of the enum?

File: include/netlink/route/link/bridge_info.h (hunk 3 above)
Comment (ronand-atl) [reply include/netlink/route/link/bridge_info.h]: Not sure I'm sold on the lack of consistency argument but I'll change it anyway

File: include/netlink/route/link/bridge_info.h (hunk 6 above)
Comment (thom311) [reply include/netlink/route/link/bridge_info.h:11]: Why do you say it won't fly?

I didn't previously realize that `BR_BOOLOPT_NO_LL_LEARN` is an enum. Often enough, such identifiers are just int (literals) in kernel header.

I don't think we should/need to use `enum br_boolopt_id`. We also avoid enums in our public API, because their size depends on compiler flags like `-fshort-enums` (not sure that's a very convincing argument).

I think just use a plain `int id`. Then also don't include `<linux/if_bridge.h>`. Let the user include that to get the enum values they need.

Comment (ronand-atl) [reply include/netlink/route/link/bridge_info.h]: > Why do you say it won't fly?

Probably because we would be including our copy of the kernel header which would inflate the header size. Also the included kernel header may be older or newer than the one on the user's system, which could be undesirable.

Changing to use an int instead.

Comment (thom311) [reply include/netlink/route/link/bridge_info.h]: >  It would use the kernel header that the user has on their system, not the one included in libnl.

Right. The one in libnl under `include/linux-private/` is only to build libnl itself. Those headers are not for the users of the library.

> That means you can't rely on the kernel header file defining a symbol that is used in the libnl header file.

That probably would be fine. We rely on kernel headers to be stable and not break API. In this case, `enum br_boolopt_id` was added in linux v5.0 (March 2019), so if we use the symbol, we would add a dependency on 5.0+ (which is old enough to be OK).

However, I still think using a plain "int" is preferable. Users who want to use this API should include `<linux/if_bridge.h>` themselves and use enums like `BR_BOOLOPT_NO_LL_LEARN`.

**Comment by thom311** [comment]
merged.

Very nice, thank you!
